### PR TITLE
[QT5] Support optional params using QVariant

### PR DESCRIPTION
**Pull request by xconverge**
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

Addresses #8648 in a non-breaking way. Uses QVariant for optional params so that a null can be passed in for optional params.

There is some more discussion and also a version using std::optional (would be a breaking change) on the issue.

I am open to any suggestions and ideas here now that I am a bit more familiar with the codebase

cc @ravinikam (2017/07) @stkrwork (2017/07) @etherealjoy (2018/02) @MartinDelille (2018/03) @muttleyxd (2019/08)

### CODE REVIEW

**Comment by etherealjoy** [comment]
@xconverge 
Lots of changes, need a bit of time to check this.

**Comment by xconverge** [comment]
> 
> 
> @xconverge
> Lots of changes, need a bit of time to check this.

Yep, no rush. The biggest change is really just losing the type info in the function definition for an optional parameter. The rest is whitespace, null checking, indentation, and using the .value of the QVariant. I opted to not use the toString functionality of QVariant yet to keep things closer to the way they were. Make sure you expand PFXPetApi.cpp since that has optional parmeters

**Comment by xconverge** [comment]
Now that I have thought about it a bit more, I like how transparent this change is vs std:optional<type> way, even though I personally can use C++17 and std:optional as a breaking change.

File: modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache [hunk 1]
```diff
@@ -226,220 +226,231 @@ void {{classname}}::{{nickname}}({{#allParams}}const {{{dataType}}} &{{paramName
...
-                fullPath.append(QUrl::toPercentEncoding(::{{cppNamespace}}::toStringValue(t)));
-                count++;
-            }
-              fullPath.replace({{paramName}}PathParam, paramString);
-    }
-{{/collectionFormat}}{{/pathParams}}
+                fullPath.replace({{paramName}}PathParam, paramString);
+        }
+    {{/collectionFormat}}
+}
```
Comment (etherealjoy) [inline modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache:304]: indentation

**Review by etherealjoy — CHANGES_REQUESTED** [review]
Seem fine to me.
Please uncomment the code for the functions below in PetApiTests.cpp, update the API usage and compile. 

```C++
    void updatePetTest();
    void updatePetWithFormTest();
```
You can leave them commented before checking in.

**Comment by etherealjoy** [comment]
Thanks for the PR, looks like a good feature increment

File: modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache (hunk 1 above)
Comment (xconverge) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache:304]: yea this was actually weirdly intentional

see that samples/client/petstore/cpp-qt5/client/PFXStoreApi.cpp and it is formatted correctly (with this indented once, it gets double indented in the output)

I will revisit it and see if there is something I am missing/a way of using what was there before to help format it right

**Comment by xconverge** [comment]
> 
> 
> Seem fine to me.
> Please uncomment the code for the functions below in PetApiTests.cpp, update the API usage and compile.
> 
> ```c++
>     void updatePetTest();
>     void updatePetWithFormTest();
> ```
> 
> You can leave them commented before checking in.

Will do, is this just for my use/verification or do you expect to see samples changed as well in the final diff?

File: modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache (hunk 1 above)
Comment (xconverge) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache]: I was able to clean it up a bit

**Comment by xconverge** [comment]
> Seem fine to me.
> Please uncomment the code for the functions below in PetApiTests.cpp, update the API usage and compile.
> 
> ```c++
>     void updatePetTest();
>     void updatePetWithFormTest();
> ```
> 
> You can leave them commented before checking in.

There were no changes necessary when uncommenting this section, and it compiles and the tests pass

File: samples/client/petstore/cpp-qt5/client/PFXPetApi.cpp [hunk 2]
```diff
@@ -333,87 +336,89 @@ void PFXPetApi::findPetsByStatus(const QList<QString> &status) {
...
+                    if (fullPath.indexOf("?") > 0)
+                        fullPath.append(queryPrefix);
+                    else
+                        fullPath.append("?");
+                    fullPath.append("status=").append(::test_namespace::toStringValue(t));
+                }
+            } else if (QString("csv").indexOf("ssv") == 0) {
                 if (fullPath.indexOf("?") > 0)
-                    fullPath.append(queryPrefix);
+                    fullPath.append("&");
```
Comment (xconverge) [inline samples/client/petstore/cpp-qt5/client/PFXPetApi.cpp:358]: this looks off

Comment (xconverge) [reply samples/client/petstore/cpp-qt5/client/PFXPetApi.cpp:358]: nevermind, just bad side by side parsing with the bracket changes

**Comment by etherealjoy** [comment]
Code does not compile
```
/home/dev/openapi-generator/samples/client/petstore/cpp-qt5/PetStore/PetApiTests.cpp: 
In member function ‘void PetApiTests::updatePetWithFormTest()’:
/home/dev/openapi-generator/samples/client/petstore/cpp-qt5/PetStore/PetApiTests.cpp:
211:44: error: conversion from ‘std::nullptr_t’ to ‘const QVariant’ is ambiguous
  211 |     api.updatePetWithForm(id, name, nullptr);

```

**Comment by xconverge** [comment]
> Code does not compile
> 
> ```
> /home/dev/openapi-generator/samples/client/petstore/cpp-qt5/PetStore/PetApiTests.cpp: 
> In member function ‘void PetApiTests::updatePetWithFormTest()’:
> /home/dev/openapi-generator/samples/client/petstore/cpp-qt5/PetStore/PetApiTests.cpp:
> 211:44: error: conversion from ‘std::nullptr_t’ to ‘const QVariant’ is ambiguous
>   211 |     api.updatePetWithForm(id, name, nullptr);
> ```
> 
> https://github.com/xconverge/openapi-generator/blob/supportOptionalParamsQt5QVariant/samples/client/petstore/cpp-qt5/PetStore/PetApiTests.cpp#L238

What command do I run to compile this? I only know of building samples and then mvn clean package or ./run-in-docker.sh clean package so far...Having trouble finding it in the CONTRIBUTING.md

**Comment by MartinDelille** [comment]
There is indeed some documentation lacking.

```
cd samples/client/petstore/cpp-qt5
mkdir build
cd build
cmake ..
make
```

**Comment by xconverge** [comment]
> 
> 
> There is indeed some documentation lacking.
> 
> ```
> cd samples/client/petstore/cpp-qt5
> mkdir build
> cd build
> cmake ..
> make
> ```
> 
> We could maybe drop the [build-and-test.bash](https://github.com/OpenAPITools/openapi-generator/blob/master/samples/client/petstore/cpp-qt5/build-and-test.bash) script memory leaks check and restore its execution in the test ci job.

Thanks, @etherealjoy sorry you had to hold my hand but I got there!

**Review by etherealjoy — APPROVED** [review]

**Comment by etherealjoy** [comment]
Unfortunately the tests were commented out due to a failure in the CI. But normally they should be compilable out of the box.

@xconverge 
I have one just one question now, if we have an API call with multiple optionals in different param order then I suppose we have to fill the others too right, cannot skip the one in between, anyway this is a C++ limitation I think.

**Comment by xconverge** [comment]
> 
> 
> Unfortunately the tests were commented out due to a failure in the CI. But normally they should be compilable out of the box.
> 
> @xconverge
> I have one just one question now, if we have an API call with multiple optionals in different param order then I suppose we have to fill the others too right, cannot skip the one in between, anyway this is a C++ limitation I think.

Good question, you can pass in QVariant() to the middle parameters, not ideal but still would "work"

Yea, just a c++ limitation

**Comment by etherealjoy** [comment]
Ok fine, I wait for the others if they have some comment to add.

**Comment by MartinDelille** [comment]
I tried to change the `required` value to `false` for the third parameter of the `updatePetWithForm` operation (https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/2_0/petstore.yaml#L202) to see what happened if the second argument as a default value but not the third and hopefully the parameters are sorted by *requirability*! 😄 

Great job!

**Review by MartinDelille — APPROVED** [review]

**Comment by xconverge** [comment]
> I tried to change the `required` value to `false` for the third parameter of the `updatePetWithForm` operation (https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/2_0/petstore.yaml#L202) to see what happened if the second argument as a default value but not the third and hopefully the parameters are sorted by _requirability_! smile
> 
> Great job!

sortModelPropertiesByRequiredFlag being defaulted to true did the legwork for that!

**Comment by wing328** [comment]
@xconverge thanks for the PR, which has been included in the v5.1.0 release: https://twitter.com/oas_generator/status/1373636654024380423

**Pull request merged**